benches/end2end.rs

```diff
@@ -0,0 +1,330 @@
+
+//         let pp = fcomm::public_params(reduction_count).unwrap();
+//         let frames = prover
+//             .get_evaluation_frames(ptr, env, &mut store, limit, &lang_vesta)
+//             .unwrap();
+
+//         b.iter(|| {
+//             let result = prover
+//                 .prove(&pp, frames.clone(), &mut store, &lang_vesta)
+//                 .unwrap();
+//             black_box(result);
+//         })
+//     });
+// }
+
+fn prove_benchmark(c: &mut Criterion) {
+    let mut group = c.benchmark_group("prove_benchmark");
+    group
+        .sampling_mode(SamplingMode::Flat)
+        .measurement_time(Duration::from_secs(120))
+        .sample_size(10);
+
+    let limit = 1_000_000_000;
+    let lang_vesta = Lang::<pasta_curves::Fq, Coproc<pasta_curves::Fq>>::new();
+    let mut store = Store::default();
+    let reduction_count = DEFAULT_REDUCTION_COUNT;
+
+    let size = (10, 0);
+    let benchmark_id = BenchmarkId::new("prove_go_base_nova", format!("_{}_{}", size.0, size.1));
+
+    group.bench_with_input(benchmark_id, &size, |b, &s| {
+        let ptr = go_base::<pasta_curves::Fq>(&mut store, s.0, s.1);
+        let prover = NovaProver::new(reduction_count, lang_vesta.clone());
+        let pp = fcomm::public_params(reduction_count).unwrap();
+        let frames = prover
+            .get_evaluation_frames(ptr, empty_sym_env(&store), &mut store, limit, &lang_vesta)
+            .unwrap();
+
+        b.iter(|| {
+            let result = prover
+                .prove(&pp, frames.clone(), &mut store, &lang_vesta)
+                .unwrap();
+            black_box(result);
+        })
+    });
+}
+
+fn verify_benchmark(c: &mut Criterion) {
+    let mut group = c.benchmark_group("verify_benchmark");
+    group
+        .measurement_time(Duration::from_secs(5))
+        .sample_size(60);
+
+    let limit = 1_000_000_000;
+    let lang_vesta = Lang::<pasta_curves::Fq, Coproc<pasta_curves::Fq>>::new();
+    let mut store = Store::default();
+    let reduction_count = DEFAULT_REDUCTION_COUNT;
+
+    let sizes = vec![(10, 16), (10, 160)];
+    for size in sizes {
+        let parameter_string = format!("_{}_{}", size.0, size.1);
+        let benchmark_id = BenchmarkId::new("verify_go_base_nova", &parameter_string);
+        group.bench_with_input(benchmark_id, &size, |b, &s| {
+            let ptr = go_base(&mut store, s.0, s.1);
+            let prover = NovaProver::new(reduction_count, lang_vesta.clone());
+            let pp = fcomm::public_params(reduction_count).unwrap();
+            let frames = prover
+                .get_evaluation_frames(ptr, empty_sym_env(&store), &mut store, limit, &lang_vesta)
+                .unwrap();
+            let proof = prover
+                .prove(&pp, frames.clone(), &mut store, &lang_vesta)
+                .unwrap();
+
+            b.iter(|| {
+                let result = proof
+                    .0
+                    .verify(&pp, proof.3, proof.1.clone(), &proof.2[..])
```

Comment:
Is this `.clone()` here a red flag?

Reply:
Ah, yes it is: the iterations consume the proof you just generated.
What you want to do is use `iter_batched` instead, and use its `setup` parameter to clone z0.

https://docs.rs/criterion/latest/criterion/struct.Bencher.html#method.iter_batched